internal/plans/deferring/deferred.go

```diff
@@ -0,0 +1,322 @@
+// calling this function.
+func NewDeferred(resourceGraph addrs.DirectedGraph[addrs.ConfigResource]) *Deferred {
+	return &Deferred{
+		resourceGraph:                    resourceGraph,
+		resourceInstancesDeferred:        addrs.MakeMap[addrs.ConfigResource, addrs.Map[addrs.AbsResourceInstance, deferredResourceInstance]](),
+		partialExpandedResourcesDeferred: addrs.MakeMap[addrs.ConfigResource, addrs.Map[addrs.PartialExpandedResource, deferredPartialExpandedResource]](),
+		partialExpandedModulesDeferred:   addrs.MakeSet[addrs.PartialExpandedModule](),
+	}
+}
+
+// SetExternalDependencyDeferred modifies a freshly-constructed [Deferred]
+// so that it will consider all resource instances as needing their actions
+// deferred, even if there's no other reason to do that.
+//
+// This must be called zero or one times before any other use of the receiver.
+// Changing this setting after a [Deferred] has already been used, or
+// concurrently with any other method call, will cause inconsistent and
+// undefined behavior.
+func (d *Deferred) SetExternalDependencyDeferred() {
+	d.externalDependencyDeferred = true
+}
+
+// HaveAnyDeferrals returns true if at least one deferral has been registered
+// with the receiver.
+//
+// This method is intended as a summary result to propagate to the modules
+// runtime caller so it can know if it should treat any downstream objects
+// as having their own changes deferred without having to duplicate the
+// modules runtime's rules for what counts as a deferral.
+func (d *Deferred) HaveAnyDeferrals() bool {
+	return d.externalDependencyDeferred ||
+		d.resourceInstancesDeferred.Len() != 0 ||
+		d.partialExpandedResourcesDeferred.Len() != 0 ||
+		len(d.partialExpandedModulesDeferred) != 0
+}
+
+// ShouldDeferResourceChanges returns true if the receiver knows some reason
+// why the resource instance with the given address should have its planned
+// action deferred for a future plan/apply round.
+//
+// This method is specifically for resource instances whose full address is
+// known and thus it would be possible in principle to plan changes, but we
+// still need to respect dependency ordering and so any planned changes must
+// be deferred if any upstream planned action was already deferred for
+// some reason.
+//
+// Callers who get the answer true should announce an approximation of the
+// action they would have planned to [Deferred.ReportResourceInstanceDeferred],
+// but should skip writing that change into the live plan so that downstream
+// evaluation will be based on the prior state (similar to in a refresh-only
+// plan) rather than the result of the deferred action.
+//
+// It's invalid to call this method for an address that was already reported
+// as deferred using [Deferred.ReportResourceInstanceDeferred], and so this
+// method will panic in that case. Callers should always test whether a resource
+// instance action should be deferred _before_ reporting that it has been.
```

Comment:
I was a little confused by this paragraph of commentary (and the corresponding one below before the panic). I think the contract we're enforcing here is that callers are required to call these two methods in a specific order for each resource instance, at most once each:

- First call `ShouldDeferResourceInstanceChanges`
- If it returns true, then you must call `ReportResourceInstanceDeferred`

The contract is to crash upon calling these in any other order, or any more often. That's a severe restriction, which may be entirely reasonable, but it was a bit hard for me to follow and I wonder if it's a possible maintenance hazard.

Assuming that I've understood what's happening, which is quite possibly not the case, I have two ideas:

- Bundle these two methods into one, so that it's impossible to call them in the wrong order. Something like:

    ```go
    func (d *Deferred) ShouldDeferResourceInstanceChanges(addr addrs.AbsResourceInstance) bool {
        ret := d.shouldDeferResourceInstanceChanges(addr)
        if ret {
            d.reportResourceInstanceDeferred(addr)
        }
        return ret
    }
    ```

- Or, update the comments throughout to clarify the expectation that these methods are always to be called in this specific order, as a pair, and only once per addr, because we require that the graph walk only evaluates deferrals once per instance.

Reply:
I agree that this is a little confusing, and I think a big part of what makes it confusing is that many different resource instances are potentially in play, and this constraint only applies for _one specific resource instance address_. In practice then, the sequence of calls is:

1. Some upstream graph node earlier called `ReportResourceInstanceDeferred` to report that _it_ has deferred a change, regardless of why that happened.
2. The current graph node then calls `ShouldDeferResourceInstanceChanges` with its _own_ address, which causes this function to check whether that any of address's upstream dependencies were deferred. (In the case of this example, yes.)
3. The current graph node then responds to that result by calling `ReportResourceInstanceDeferred` with its _own_ resource instance address. (This would then be the step 1 of a downstream incarnation of this sequence, if any)

Bundling the check in with the report is an interesting idea. I think we will still need a separate exported `ReportResourceInstanceDeferred` because in the end this will not be the _only_ reason for a resource instance graph node to need to report that it made a deferral decision: it would do so also if a provider responded to `PlanResourceChange` by saying "I cannot plan this yet", once we get to that functionality in a later PR. But I admittedly cannot think of any reason why one would call `SourceDeferResourceInstanceChanges`, get a `true` response, and _not_ immediately call `ReportResourceInstanceDeferred`, so it does seem plausible to build it so that asking the question and performing the side-effect of reporting are an atomic action from the caller's perspective.

(I do typically try to avoid writing functions that are both accessors and mutators at the same time as far as that's possible to do so, but I do agree that in this case the separation seems more likely to encourage bugs than combining them would.)